PMNS_ScalarNSI.cxx

```diff
@@ -0,0 +1,410 @@
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_ScalarNSI::SetEta_mutau(double a, double phi){ SetEta(1,2, a, phi); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_ScalarNSI::SetEta_etau(double a, double phi){ SetEta(0,2, a, phi); }
+
+
+////Get lightest neutrino mass
+
+
+double PMNS_ScalarNSI::GetM(){ return fM;}
+/// Set eps_mutau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a   - The absolute value of the parameter eps_mutau
+///
+/// Build the full Hamiltonian in matter.
+///
+/// Here we divide the mass squared matrix Hms by the 2E
+/// to obtain the vacuum Hamiltonian in eV. Then, the matter
+/// potential is added to the electron component.
+///
+void PMNS_ScalarNSI::UpdateHam()
+{
+
+  double lv = 2 * kGeV2eV*fEnergy;     // 2E in eV 
+
+  double kr2GNe = kK2*M_SQRT2*kGf;
+  kr2GNe *= fPath.density * fPath.zoa; // Matter potential in eV
+
+  // PMNS matrix U
+  
+  complexD U[3][3];
+  int i,j,k;
+  for( i=0;i<3;i++)
+  {
+    
+     for( j=0;j<3;j++)
+     {
+        U[i][j]=fEvec[i][j];
+     }
+  }
+
+  
+
+
+
+complexD M[3][3], eta[3][3];
+
+/// Defining mass matrix M=Diag{m1,m2,m3}, and scalar NSI matrix for NO
+  if (fDm[2]>0)
+  {
+  
+  double m1=fM;
+  double m2=sqrt(fDm[1]+m1*m1);
+  double m3=sqrt(fDm[2]+m1*m1);
+  
+   
+  
+  M[0][0]=m1;
+  M[1][1]=m2;
+  M[2][2]=m3;
+  
+  
+  
+    for( i=0;i<3;i++)
+    {
+      for( j=i+1;j<3;j++)
+      {
+      M[i][j]=0;
+      }
+    }
+
+
+
+eta[0][0]=fEta[0][0]*sqrt(fDm[2]);
+eta[1][1]=fEta[1][1]*sqrt(fDm[2]);
+eta[2][2]=fEta[2][2]*sqrt(fDm[2]);
+
+ eta[0][1]=fEta[0][1]*sqrt(fDm[2]);
+eta[0][2]=fEta[0][2]*sqrt(fDm[2]);
+eta[1][2]=fEta[1][2]*sqrt(fDm[2]);
+
+
+eta[1][0]=conj(eta[0][1]);
+eta[2][0]=conj(eta[0][2]);
+eta[2][1]=conj(eta[1][2]);
+ } 
+/// Defining mass matrix M=Diag{m1,m2,m3}, and scalar NSI matrix Eta for IO
+if(fDm[2]<0)
+{
+  double m3=fM;
+  double m1=sqrt(m3*m3-fDm[2]);
+  double m2=sqrt(fDm[1]+m1*m1);
+  
+   
+  
+  M[0][0]=m1;
+  M[1][1]=m2;
+  M[2][2]=m3;
+  
+  
+  
+    for( i=0;i<3;i++)
+    {
+      for( j=i+1;j<3;j++)
+      {
+      M[i][j]=0;
+      }
+    }
+
+
+eta[0][0]=fEta[0][0]*sqrt(-fDm[2]);
+eta[1][1]=fEta[1][1]*sqrt(-fDm[2]);
+eta[2][2]=fEta[2][2]*sqrt(-fDm[2]);
+
+ eta[0][1]=fEta[0][1]*sqrt(-fDm[2]);
+eta[0][2]=fEta[0][2]*sqrt(-fDm[2]);
+eta[1][2]=fEta[1][2]*sqrt(-fDm[2]);
+
+
+eta[1][0]=conj(eta[0][1]);
+eta[2][0]=conj(eta[0][2]);
+eta[2][1]=conj(eta[1][2]);
+
+
+}
+ /// Calculating U^{\dagger}
+  complexD Udag[3][3];
+  
+  for(int i=0;i<3;i++)
+  {
+    
+    for(int j=0;j<3;j++)
+    {
+      Udag[i][j]=conj(U[j][i]);
+    }
+  }
```

Comment:
Again, why copy this when you can use the `conj` operation directly when needed? Keep in mind copying does take time.